extensions/src/test/java/org/pentaho/platform/plugin/services/metadata/PentahoAsyncThreadRunnerTest.java

```diff
@@ -0,0 +1,82 @@
+ * Foundation.
+ *
+ * You should have received a copy of the GNU Lesser General Public License along with this
+ * program; if not, you can obtain a copy at http://www.gnu.org/licenses/old-licenses/lgpl-2.1.html
+ * or from the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Lesser General Public License for more details.
+ *
+ * Copyright (c) 2022 Hitachi Vantara.  All rights reserved.
+ */
+
+package org.pentaho.platform.plugin.services.metadata;
+
+import junit.framework.TestCase;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+public class PentahoAsyncThreadRunnerTest extends TestCase {
+
+  @Test
+  public void testAsyncRun() throws Exception {
+    final ConcurrentLinkedQueue<Integer> queue = new ConcurrentLinkedQueue();
+    // SETUP
+    Collection<Integer> integers = Arrays.asList( 1, 2, 3, 4 );
+
+    Collection<Callable<Integer>> tasks = new ArrayList<>();
+    for ( Integer i : integers ) {
+      // update queue to verify later tasks actually run
+      tasks.add( () -> { queue.add( i ); return i; }  );
+    }
+
+    ExecutorService executorService = Executors.newFixedThreadPool( 1 );
+    PentahoAsyncThreadRunner patr = new PentahoAsyncThreadRunner( 1, null );
+
+    // EXECUTE
+    Thread thread = patr.asyncRun( executorService, tasks );
+
+    // VERIFY
+    assertNotNull( thread );
+    assertTrue( "Timeout happened before threads stopped",
+            executorService.awaitTermination( 5, TimeUnit.SECONDS ) ); // should complete way before this
+    integers.stream().forEach( i -> assertTrue( "thread/task " + i + " not found", queue.contains( i ) ) );
+  }
+
+  @Test
+  public void testExecuteTasks() throws Exception {
+    // SETUP
+    Callable<Integer> task1 = () -> { return 1; };
+    Callable<Integer> task2 = () -> { return null; }; // null values won't be returned
+    Callable<Integer> task3 = () -> { return 3; };
+    Callable<Integer> task4 = () -> { throw new Exception( "error in callable, but ignore" ); }; // exceptions won't be returned
+
+    Collection<Callable<Integer>> tasks = Arrays.asList( task1, task2, task3, task4 );
+
+    ExecutorService executorService = Executors.newFixedThreadPool( 1 );
+
+    PentahoAsyncThreadRunner patr = new PentahoAsyncThreadRunner( 1, null );
+
+    // EXECUTE
+    // ignore errors in console thrown from task2 and task4
+    Collection<Integer> actualExecuteTasks = patr.executeTasks( executorService, tasks );
```

Comment:
For a general class, I'm not sure if filtering out `null` results is warranted. May come as susprising.
As for errors, that's a thing, but a different one.

This reminds me of the various JavaScript static `Promise.<verb>` methods, each performing the wait on multiple promises in different ways regarding termination and errors. (e.g. [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) vs [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled).

I'm sure Java has APIs with equivalent functionality, I just don't know them well, and so cannot easily evaluate how "idiomatic" and clear the name `executeTasks` is for the type of "wait" being performed.

Reply:
I'm gonna follow the KISS principle. 

No change.

I too am a familiar with the javascript concept.

For now, this class supports the one and only caller SessionCachingMetadataDomainRepository. In the future, more functionality and more robustness can be added.

Reply:
My comment is not about making things complex or complicated, just clear and predictable (the principle of least surprise). Filtering out `null` values is something I find very context specific, and thus surprising. I would then expect that the method either doesn't do it or that it had a name that would reveal it.

Reply:
will move PentahoAsyncThreadRunnerTest.java back to inner static class of SessionCachingMetadataDomainRepository.java this will remove the need to extend the functionality of the class since only SessionCachingMetadataDomainRepository will be the client of it.

Reply:
see commit: https://github.com/pentaho/pentaho-platform/pull/5124/commits/4eb7eb5e6be8ae5e146ac8818babf2f3b5b60a20